### PR TITLE
Greedy allocator

### LINKED ISSUE #36: Greedy allocation algorithm

### LINKED ISSUE #37: Allocator.rst example should use values_count to count new occupancies
e.g.:

```python
        infrastructure.occupancy += (
            newly_assigned.value_counts(dropna=True)
            .reindex(infrastructure.index)
            .fillna(0)
        )
```

### LINKED ISSUE #42: Howto for algorithm comparison
Compare greedy and random allocator by plotting distances in violin plot.

### PR DESCRIPTION
**Pull request by mdavezac**
- Adds greedy allocator

### CODE REVIEW

**Comment by mdavezac** [comment]
@cc-a, @toytun, @DT16, this PR  implements a greedy allocator and is ready for review. Usage of the algorithm is documented, however I'll add a notebook in another PR.

**Comment by mdavezac** [comment]
mmm, okay, so I added a notebook for comparing allocators to this PR rather than a distinct one  by mistake... so it's all here.

File: docs/conf.py [hunk 1]
```diff
@@ -6,12 +6,13 @@
...
 
+
 # If extensions (or modules to document with autodoc) are in another directory,
 # add these directories to sys.path here. If the directory is relative to the
+import sys
+
 # documentation root, use os.path.abspath to make it absolute, like shown here.
-#
 from datetime import datetime  # noqa: E402
-import sys
```
Comment (cc-a) [inline docs/conf.py:16]: `isort`'s doing?

File: src/evosim/allocators.py [hunk 2]
```diff
@@ -164,7 +165,161 @@ def random_overbooking(
...
+    Returns:
+        pandas.DataFrame: a shallow copy the fleet with an extra column, "allocation",
+        indicating the label of the allocated post into the ``charging_posts`` table. Or
+        ``pd.NA`` if the vehicle is not allocated.
+    """
+    from sklearn.neighbors import BallTree
+    from evosim.matchers import match_all_to_all
+    from warnings import warn
+
+    if nearest_neighbors is None or nearest_neighbors <= 0:
```
Comment (cc-a) [inline src/evosim/allocators.py:221]: Default argument for `nearest_neighbors` is 10. Do we need a check for None?

File: src/evosim/allocators.py [hunk 3]
```diff
@@ -164,7 +165,161 @@ def random_overbooking(
...
+    if maxiter is None or maxiter <= 0:
+        maxiter = (charging_posts.capacity - charging_posts.occupancy).sum() + 1
+
+    def locations(data: pd.DataFrame, is_current=True) -> np.ndarray:
+        names = ("latitude", "longitude") if is_current else ("dest_lat", "dest_long")
+        return (
+            np.concatenate(
+                (
+                    data[names[0]].to_numpy()[:, None],
+                    data[names[1]].to_numpy()[:, None],
```
Comment (cc-a) [inline src/evosim/allocators.py]: More explicit
```suggestion
                    data[names[0]].to_numpy()[:, np.newaxis],
                    data[names[1]].to_numpy()[:, np.newaxis],
```

File: src/evosim/allocators.py [hunk 4]
```diff
@@ -164,7 +165,161 @@ def random_overbooking(
...
+    Hence shuffling the fleet may well yield different results.
+
+    Args:
+        fleet (pandas.DataFrame): table defining the fleet of vehicles
+        charging_posts (pandas.DataFrame): table defining the charging infrastructure
+        matcher: defines the compatibility between an electric vehicle and a charging
+            post
+        distance: the distance metric used to compute the nearest neighbors. See
+            :py:class:`sklearn.neighbors.BallTree`.
+        leaf_size: parameter of :py:class:`sklearn.neighbors.BallTree`
```
Comment (cc-a) [inline src/evosim/allocators.py:203]: Missing entry for nearest_neighbors

File: src/evosim/allocators.py [hunk 5]
```diff
@@ -164,7 +165,161 @@ def random_overbooking(
...
+    result["allocation"] = allocation
+    return result
+
+
+def _pick_first(group):
+    """Pick only as many as there are vacancies."""
+    nvacancies = group.vacancies.iloc[0]
+    if len(group) < nvacancies:
+        return group.allocation
+    return group.allocation.where(np.arange(len(group), dtype=int) < nvacancies, pd.NaT)
```
Comment (cc-a) [inline src/evosim/allocators.py]: Appears to be unused in favour of `pick_first`.

File: src/evosim/allocators.py [hunk 6]
```diff
@@ -164,7 +165,161 @@ def random_overbooking(
...
+
+
+def _void_overbooking(
+    fleet: pd.DataFrame, charging_posts: pd.DataFrame, allocation: pd.Series
+) -> pd.Series:
+    """Set overbooked allocations to NaT."""
```
Comment (cc-a) [inline src/evosim/allocators.py]: Is this docstring accurate?

File: src/evosim/allocators.py [hunk 7]
```diff
@@ -164,7 +165,161 @@ def random_overbooking(
...
+            post
+        distance: the distance metric used to compute the nearest neighbors. See
+            :py:class:`sklearn.neighbors.BallTree`.
+        leaf_size: parameter of :py:class:`sklearn.neighbors.BallTree`
+        maxiter: maximum number of iterations before bailing out. Defaults to the number
+            of vacancies.
+
+            .. note::
+
+                Earch turn of the loop, the iteration counter is increased by the number
```
Comment (cc-a) [inline src/evosim/allocators.py]:
```suggestion
                Each turn of the loop, the iteration counter is increased by the number
```

File: src/evosim/allocators.rst [hunk 8]
```diff
@@ -73,100 +42,221 @@ follows:
...
+           [9, 3, 5, 6, 1, 0, 4, 8, 2, 7],
+           [2, 1, 7, 8, 5, 4, 3, 9, 6, 0]]...)
+
+The first row of the matrix above corresponds to the first electric vehicle. It gives
+the indices (as in :py:meth:`pandas.DataFrame.iloc`, not the labels of
+:py:meth:`pandas.DataFrame.loc`) of the posts, left to right from nearest to furthest.
+The distance matrix follows the same structure.
+
+
+.. topic:: Distances on the surface of the Eartch
```
Comment (cc-a) [inline src/evosim/allocators.rst:204]:
```suggestion
.. topic:: Distances on the surface of the Earth
```

**Review by cc-a — APPROVED** [review]

File: docs/conf.py (hunk 1 above)
Comment (mdavezac) [reply docs/conf.py:16]: yeah

File: src/evosim/allocators.py (hunk 2 above)
Comment (mdavezac) [reply src/evosim/allocators.py:221]: it does already check for none

**Pull request merged**